crates/dbsp/src/trace/spine_async/mod.rs

```diff
@@ -83,81 +76,360 @@ impl<B: Batch + Send + Sync> From<&Spine<B>> for CommittedSpine<B> {
+    /// them into merging batches, and returns those batches. Otherwise, returns
+    /// `None` without changing anything.
+    fn try_start_merge(&mut self) -> Option<[Arc<B>; 2]> {
+        if self.merging_batches.is_none() && self.loose_batches.len() >= 2 {
+            let a = self.loose_batches.pop().unwrap();
+            let b = self.loose_batches.pop().unwrap();
+            self.merging_batches = Some([Arc::clone(&a), Arc::clone(&b)]);
+            Some([a, b])
+        } else {
+            None
+        }
+    }
+}
+
+/// State shared between the merger thread and the main thread.
 ///
-/// If there's only one batch then the `Arc` has a reference count of 1,
-/// otherwise they are higher than that because the background merger thread
-/// holds a reference.
-#[derive(SizeOf)]
-pub struct BatchState<B>(Vec<Arc<B>>)
+/// This shared state is accessed through a `Mutex`, which we try to hold for as
+/// short a time as possible.
+struct SharedState<B>
 where
-    B: Batch;
+    B: Batch,
+{
+    key_filter: Option<Filter<B::Key>>,
+    value_filter: Option<Filter<B::Val>>,
+    slots: [Slot<B>; MAX_LEVELS],
+    request_exit: bool,
+    merge_stats: MergeStats,
+}
 
-impl<B> BatchState<B>
+impl<B> SharedState<B>
 where
     B: Batch,
 {
-    /// The number of actual updates contained in the level.
-    fn n_updates(&self) -> usize {
-        self.0.iter().map(|b| b.len()).sum()
+    pub fn new() -> Self {
+        Self {
+            key_filter: None,
+            value_filter: None,
+            slots: std::array::from_fn(|_| Slot::default()),
+            request_exit: false,
+            merge_stats: MergeStats::default(),
+        }
+    }
+
+    /// Adds all of `batches` as (initially) loose batches.  They will be merged
+    /// when the merger thread has a chance (although it might not be awake).
+    fn add_batches(&mut self, batches: impl IntoIterator<Item = Arc<B>>) {
+        for batch in batches {
+            if !batch.is_empty() {
+                self.slots[Spine::<B>::size_to_level(batch.len())]
+                    .loose_batches
+                    .push(batch);
+            }
+        }
     }
 
-    fn batch_count(&self) -> usize {
-        self.0.len()
+    /// Add `batches` as an (initially) loose batch, which will be merged when
+    /// the merger thread has a chance (although it might not be awake).
+    ///
+    /// Returns a copy of all of the batches (whether loose or being merged).
+    fn add_batch_and_get_batches(&mut self, batch: Arc<B>) -> Vec<Arc<B>> {
+        self.add_batches([batch]);
+        self.get_batches()
     }
 
+    fn get_filters(&self) -> (Option<Filter<B::Key>>, Option<Filter<B::Val>>) {
+        (self.key_filter.clone(), self.value_filter.clone())
+    }
+
+    /// Gets a copy of all of the batches (whether loose or being merged).
+    fn get_batches(&self) -> Vec<Arc<B>> {
+        let mut batches = Vec::new();
```

Comment:
can you use with_capacity here since you know how many batches you have

Reply:
Done.